### PR TITLE
GODRIVER-1925 Surface cursor errors in DownloadStream fillBuffer

### PR DESCRIPTION
**Pull request by benjirewis**
GODRIVER-1925

Checks for a possible cursor error in `DownloadStream#FillBuffer` instead of always returning `errNoMoreChunks` when `ds.cursor.Next` returns false. If there is a cursor error, closes the cursor and returns the underlying error.

Adds a test to simulate a timeout during `DownloadStream#Read` and assert that `context deadline exceeded` is returned instead of `io.EOF`.

### CODE REVIEW

File: mongo/gridfs/download_stream.go [hunk 1]
```diff
@@ -226,6 +226,11 @@ func (ds *DownloadStream) GetFile() *File {
 func (ds *DownloadStream) fillBuffer(ctx context.Context) error {
 	if !ds.cursor.Next(ctx) {
 		ds.done = true
+		// Check for cursor error, otherwise there are no more chunks.
+		if ds.cursor.Err() != nil {
+			ds.cursor.Close(ctx)
```
Comment (benjirewis) [inline mongo/gridfs/download_stream.go:231]: `Close` can return an error, but I'm not sure what to do with it here. Should it supersede the `ds.cursor.Err()`?

File: mongo/integration/gridfs_test.go [hunk 2]
```diff
@@ -369,6 +370,40 @@ func TestGridFS(x *testing.T) {
...
+			// set in the past, this should cause a timeout.
+
+			// Create file to upload.
+			fileName := "read-error-test"
+			fileData := make([]byte, 17000000)
+
+			// Create bucket.
+			bucket, err := gridfs.NewBucket(mt.DB)
+			assert.Nil(mt, err, "NewBucket error: %v", err)
+			defer func() { _ = bucket.Drop() }()
```
Comment (matthewdale) [inline mongo/integration/gridfs_test.go:385]: What's the purpose of the ` _ = ` in this `bucket.Drop()` statement? I see it in other tests, too.

**Review by matthewdale — APPROVED** [review]
👍

File: mongo/integration/gridfs_test.go (hunk 2 above)
Comment (benjirewis) [reply mongo/integration/gridfs_test.go:385]: `bucket.Drop()` drops the "files" and "chunks" collections associated with the bucket. In testing code, we generally explicitly ignore the potential error in a deferred collection drop (or client disconnect, etc.) with `_ =`.  I suppose it's clearer than just `bucket.Drop()`, and a potential error from collection dropping is of little relevance to the test. Anyone else have context on this, though?

Comment (divjotarora) [reply mongo/integration/gridfs_test.go:385]: It's to prevent errors from `errcheck`. The other option is to modify the [errcheck-excludes](https://github.com/mongodb/mongo-go-driver/blob/master/.errcheck-excludes) file to indicate that errors from `Drop` can safely be ignored, but that takes effect globally, so we try not to use it too often.

**Review by iwysiu — APPROVED** [review]
lgtm! though as a nit, I don't feel like a lot of the comments in gridfs_test.go are necessary

File: mongo/gridfs/download_stream.go (hunk 1 above)
Comment (divjotarora) [reply mongo/gridfs/download_stream.go:231]: I agree with your decision to ignore the error from `Close` here because we want to surface the error that was actually related to iteration. Let's make it more explicit by doing `_ = ds.cursor.Close()`, though.

File: mongo/integration/gridfs_test.go [hunk 3]
```diff
@@ -369,6 +370,40 @@ func TestGridFS(x *testing.T) {
...
+			// Open a download stream to fileName.
+			ds, err := bucket.OpenDownloadStreamByName(fileName)
+			assert.Nil(mt, err, "OpenDownloadStreamByName error: %v", err)
+
+			// Set a read deadline in the past to cause a timeout.
+			err = ds.SetReadDeadline(time.Now().Add(-1 * time.Second))
+			assert.Nil(mt, err, "SetReadDeadline error: %v", err)
+
+			// Try to read.
+			p := make([]byte, 17000000)
```
Comment (divjotarora) [inline mongo/integration/gridfs_test.go:401]: nit: Rather than reusing the 17MB constant, you can use `len(fileData)` here.

**Review by divjotarora — COMMENTED** [review]
Should we also close `ds.cursor` in `DownloadStream#Close`? If a user creates a `DownloadStream` and then closes it before fully reading the file, we should close the cursor to prevent cursor/session leaks in the application.

File: mongo/gridfs/download_stream.go (hunk 1 above)
Comment (benjirewis) [reply mongo/gridfs/download_stream.go:231]: Done.

File: mongo/integration/gridfs_test.go (hunk 3 above)
Comment (benjirewis) [reply mongo/integration/gridfs_test.go:401]: Done.

**Review by benjirewis — COMMENTED** [review]
Removed some extraneous comments from the gridfs test and `Close`d the underlying cursor in `DownloadStream#Close()`.

File: mongo/gridfs/download_stream.go [hunk 4]
```diff
@@ -118,8 +118,11 @@ func (ds *DownloadStream) Close() error {
 		return ErrStreamClosed
 	}
 
-	ds.closed = true
-	return nil
+	err := ds.cursor.Close(context.Background())
+	if err != nil {
+		ds.closed = true
```
Comment (divjotarora) [inline mongo/gridfs/download_stream.go:123]: Why are we only setting `closed` to true if the error was non-nil?

Comment (benjirewis) [reply mongo/gridfs/download_stream.go:123]: I assumed if there was an error during Closing, we would not be sure the cursor was actually `closed`.

Comment (divjotarora) [reply mongo/gridfs/download_stream.go:123]: Closing the cursor is a best-effort attempt. At the very least, `ds.cursor.Close()` will release the implicit session that's being held by the cursor. If there's an error running the actual `killCursors` command, we should still set `ds.closed` to true because we tried and there's not really anything else we can do at that point. The cursor will be reaped on the server once it times out.

Comment (benjirewis) [reply mongo/gridfs/download_stream.go:123]: Yep, that sounds good. Thanks for the explanation; I had forgotten about reaping on the server! I do have to check if `ds.cursor` is nil before I close it, as there is a spec test where download length is zero.

**Review by divjotarora — APPROVED** [review]

File: mongo/gridfs/download_stream.go [hunk 5]
```diff
@@ -119,6 +119,9 @@ func (ds *DownloadStream) Close() error {
 	}
 
 	ds.closed = true
+	if ds.cursor != nil {
+		return ds.cursor.Close(context.Background())
```
Comment (kevinAlbs) [inline mongo/gridfs/download_stream.go:123]: No need to change. I was initially slightly concerned. This adds blocking work with `context.Background()`, which could block indefinitely. But previously this would always return immediately.

Though, if the cursor is exhausted, this should do no work (killCursors won't be sent to the server). If the cursor is not exhausted, this still seems preferable to leaking cursors server-side in the event of an error.

File: mongo/integration/gridfs_test.go [hunk 6]
```diff
@@ -369,6 +370,34 @@ func TestGridFS(x *testing.T) {
 			_, err = bucket.OpenDownloadStream(oid)
 			assert.Equal(mt, gridfs.ErrMissingChunkSize, err, "expected error %v, got %v", gridfs.ErrMissingChunkSize, err)
 		})
+		mt.Run("cursor error during read after downloading", func(mt *mtest.T) {
+			// To simulate a cursor error we upload a file larger than the 16MB default batch size,
+			// so Read() will actually execute a getMore against the server. Since the ReadDeadline is
```
Comment (kevinAlbs) [inline mongo/integration/gridfs_test.go:375]: IIUC the test itself won't send a `getMore`. It should fail on the first iteration of the cursor.
```suggestion
			// so the underlying cursor remains open on the server. Since the ReadDeadline is
```

File: mongo/integration/gridfs_test.go [hunk 7]
```diff
@@ -369,6 +370,34 @@ func TestGridFS(x *testing.T) {
...
+			// set in the past, this should cause a timeout.
```
Comment (kevinAlbs) [inline mongo/integration/gridfs_test.go:376]: Clever!

File: mongo/integration/gridfs_test.go [hunk 8]
```diff
@@ -369,6 +370,34 @@ func TestGridFS(x *testing.T) {
...
+			assert.Nil(mt, err, "UploadFromStream error: %v", err)
+
+			ds, err := bucket.OpenDownloadStreamByName(fileName)
+			assert.Nil(mt, err, "OpenDownloadStreamByName error: %v", err)
+
+			err = ds.SetReadDeadline(time.Now().Add(-1 * time.Second))
+			assert.Nil(mt, err, "SetReadDeadline error: %v", err)
+
+			p := make([]byte, len(fileData))
+			_, err = ds.Read(p)
```
Comment (kevinAlbs) [inline mongo/integration/gridfs_test.go:396]: Can you parameterize this test case, or add a test for `Skip` as well? This reminded me about GODRIVER-1806, and how `Read` and `Skip` are near duplicates, but `Skip` avoids extra copies.

**Review by kevinAlbs — APPROVED** [review]
LGTM with an extra test case for `Skip`!

File: mongo/gridfs/download_stream.go (hunk 5 above)
Comment (benjirewis) [reply mongo/gridfs/download_stream.go:123]: A good point and valid concern. But, leaking cursors doesn't seem great; and on the driver side, not closing the cursor in this case will leave a session open (`client.NumberSessionsInProgress() > 0`).

File: mongo/integration/gridfs_test.go (hunk 6 above)
Comment (benjirewis) [reply mongo/integration/gridfs_test.go:375]: Ah you're right. Previously I was using a `failpoint` on `getMore`, but the timeout error just comes from iterating the cursor. Changed.

File: mongo/integration/gridfs_test.go (hunk 8 above)
Comment (benjirewis) [reply mongo/integration/gridfs_test.go:396]: Ah great catch. `Read` and `Skip` both call `fillBuffer`, where the bug in error surfacing was. Added a different test case for `Skip` (wasn't sure how to parameterize given `Skip` and `Read` have different function signatures).
